README.md
# sehir-deep-learning-course

assignment-3/history.txt
History (from Greek ἱστορία, historia, meaning "inquiry, knowledge acquired by investigation")[2] is the study of the past as it is described in written documents.[3][4] Events occurring before written record are considered prehistory. It is an umbrella term that relates to past events as well as the memory, discovery, collection, organization, presentation, and interpretation of information about these events. Scholars who write about history are called historians.

History can also refer to the academic discipline which uses a narrative to examine and analyse a sequence of past events, and objectively determine the patterns of cause and effect that determine them.[5][6] Historians sometimes debate the nature of history and its usefulness by discussing the study of the discipline as an end in itself and as a way of providing "perspective" on the problems of the present.[5][7][8][9]

Stories common to a particular culture, but not supported by external sources (such as the tales surrounding King Arthur), are usually classified as cultural heritage or legends, because they do not show the "disinterested investigation" required of the discipline of history.[10][11] Herodotus, a 5th-century BC Greek historian is considered within the Western tradition to be the "father of history", and, along with his contemporary Thucydides, helped form the foundations for the modern study of human history. Their works continue to be read today, and the gap between the culture-focused Herodotus and the military-focused Thucydides remains a point of contention or approach in modern historical writing. In Asia, a state chronicle, the Spring and Autumn Annals was known to be compiled from as early as 722 BC although only 2nd-century BC texts survived.

Ancient influences have helped spawn variant interpretations of the nature of history which have evolved over the centuries and continue to change today. The modern study of history is wide-ranging, and includes the study of specific regions and the study of certain topical or thematical elements of historical investigation. Often history is taught as part of primary and secondary education, and the academic study of history is a major discipline in university studies.

The word history comes ultimately from Ancient Greek ἱστορία[12] (historía), meaning "inquiry","knowledge from inquiry", or "judge". It was in that sense that Aristotle used the word in his Περὶ Τὰ Ζῷα Ἱστορίαι[13] (Perì Tà Zôa Ηistoríai "Inquiries about Animals"). The ancestor word ἵστωρ is attested early on in Homeric Hymns, Heraclitus, the Athenian ephebes' oath, and in Boiotic inscriptions (in a legal sense, either "judge" or "witness", or similar).

The Greek word was borrowed into Classical Latin as historia, meaning 'investigation, inquiry, research, account, description, written account of past events, writing of history, historical narrative, recorded knowledge of past events, story, narrative'. History was borrowed from Latin (possibly via Old Irish or Old Welsh) into Old English as stær ('history, narrative, story'), but this word fell out of use in the late Old English period.[14]

Meanwhile, as Latin became Old French (and Anglo-Norman), historia developed into forms such as istorie, estoire, and historie, with new developments in the meaning: 'account of the events of a person's life (beginning of the 12th century), chronicle, account of events as relevant to a group of people or people in general (1155), dramatic or pictorial representation of historical events (c. 1240), body of knowledge relative to human evolution, science (c. 1265), narrative of real or imaginary events, story (c. 1462)'.[14]

It was from Anglo-Norman that history was borrowed into Middle English, and this time the loan stuck. It appears in the thirteenth-century Ancrene Wisse, but seems to have become a common word in the late fourteenth century, with an early attestation appearing in John Gower's Confessio Amantis of the 1390s (VI.1383): 'I finde in a bok compiled | To this matiere an old histoire, | The which comth nou to mi memoire'. In Middle English, the meaning of history was "story" in general. The restriction to the meaning 'the branch of knowledge that deals with past events; the formal record or study of past events, esp. human affairs' arose in the mid-fifteenth century.[14]

With the Renaissance, older senses of the word were revived, and it was in the Greek sense that Francis Bacon used the term in the late sixteenth century, when he wrote about "Natural History". For him, historia was "the knowledge of objects determined by space and time", that sort of knowledge provided by memory (while science was provided by reason, and poetry was provided by fantasy).[15]

In an expression of the linguistic synthetic vs. analytic/isolating dichotomy, English like Chinese (史 vs. 诌) now designates separate words for human history and storytelling in general. In modern German, French, and most Germanic and Romance languages, which are solidly synthetic and highly inflected, the same word is still used to mean both "history" and "story".

The adjective historical is attested from 1661, and historic from 1669.[16]

Historian in the sense of a "researcher of history" is attested from 1531. In all European languages, the substantive "history" is still used to mean both "what happened with men", and "the scholarly study of the happened", the latter sense sometimes distinguished with a capital letter, "History", or the word historiography.[13]

The title page to The Historians' History of the World

assignment-3/main.py
import sys
import numpy
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.layers import LSTM
from keras.callbacks import ModelCheckpoint
from keras.utils import np_utils

# load ascii text and covert to lowercase
filename = "history.txt"
raw_text = open(filename).read()
raw_text = raw_text.lower()

# create mapping of unique chars to integers, and a reverse mapping
chars = sorted(list(set(raw_text)))
char_to_int = dict((c, i) for i, c in enumerate(chars))
int_to_char = dict((i, c) for i, c in enumerate(chars))

# summarize the loaded data
n_chars = len(raw_text)
n_vocab = len(chars)
print "Total Characters: ", n_chars
print "Total Vocab: ", n_vocab

# prepare the dataset of input to output pairs encoded as integers
seq_length = 100
dataX = []
dataY = []

for i in range(0, n_chars - seq_length, 1):
    seq_in = raw_text[i:i + seq_length]
    seq_out = raw_text[i + seq_length]
    dataX.append([char_to_int[char] for char in seq_in])
    dataY.append(char_to_int[seq_out])
n_patterns = len(dataX)
print "Total Patterns: ", n_patterns

# reshape X to be [samples, time steps, features]
X = numpy.reshape(dataX, (n_patterns, seq_length, 1))

# normalize
X = X / float(n_vocab)

# one hot encode the output variable
y = np_utils.to_categorical(dataY)

# define the LSTM model
model = Sequential()
model.add(
    LSTM(256, input_shape=(X.shape[1], X.shape[2]), return_sequences=True))
model.add(Dropout(0.2))
model.add(LSTM(256))
model.add(Dropout(0.2))
model.add(Dense(y.shape[1], activation='softmax'))

# load the network weights
filename = "weights-improvement-02-3.0915-bigger.hdf5"
model.load_weights(filename)
model.compile(loss='categorical_crossentropy', optimizer='adam')

# pick a random seed
start = numpy.random.randint(0, len(dataX) - 1)
pattern = dataX[start]
print "Seed:"
print "\"", ''.join([int_to_char[value] for value in pattern]), "\""

# generate characters
for i in range(1000):
    x = numpy.reshape(pattern, (1, len(pattern), 1))
    x = x / float(n_vocab)
    prediction = model.predict(x, verbose=0)
    index = numpy.argmax(prediction)
    result = int_to_char[index]
    seq_in = [int_to_char[value] for value in pattern]
    sys.stdout.write(result)
    pattern.append(index)
    pattern = pattern[1:len(pattern)]

print "\nDone."


assignment-3/train.py
# Larger LSTM Network to Generate Text for Alice in Wonderland
import numpy
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.layers import LSTM
from keras.callbacks import ModelCheckpoint
from keras.utils import np_utils

# load ascii text and covert to lowercase
filename = "history.txt"
raw_text = open(filename).read()
raw_text = raw_text.lower()

# create mapping of unique chars to integers
chars = sorted(list(set(raw_text)))
char_to_int = dict((c, i) for i, c in enumerate(chars))

# summarize the loaded data
n_chars = len(raw_text)
n_vocab = len(chars)
print "Total Characters: ", n_chars
print "Total Vocab: ", n_vocab

# prepare the dataset of input to output pairs encoded as integers
seq_length = 100
dataX = []
dataY = []
for i in range(0, n_chars - seq_length, 1):
    seq_in = raw_text[i:i + seq_length]
    seq_out = raw_text[i + seq_length]
    dataX.append([char_to_int[char] for char in seq_in])
    dataY.append(char_to_int[seq_out])
n_patterns = len(dataX)
print "Total Patterns: ", n_patterns

# reshape X to be [samples, time steps, features]
X = numpy.reshape(dataX, (n_patterns, seq_length, 1))

# normalize
X = X / float(n_vocab)

# one hot encode the output variable
y = np_utils.to_categorical(dataY)

# define the LSTM model
model = Sequential()
model.add(
    LSTM(256, input_shape=(X.shape[1], X.shape[2]), return_sequences=True))
model.add(Dropout(0.2))
model.add(LSTM(256))
model.add(Dropout(0.2))
model.add(Dense(y.shape[1], activation='softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam')

# define the checkpoint
filepath = "weights-improvement-{epoch:02d}-{loss:.4f}-bigger.hdf5"
checkpoint = ModelCheckpoint(
    filepath, monitor='loss', verbose=1, save_best_only=True, mode='min')
callbacks_list = [checkpoint]

# fit the model
model.fit(X, y, epochs=50, batch_size=64, callbacks=callbacks_list)


assignment-3/antoloji.py
import scrapy


class Antoloji(scrapy.Spider):
    name = 'Antoloji'
    url = 'https://www.antoloji.com'
    poet = 'oktay-rifat'

    def start_requests(self):
        yield scrapy.Request('%s/%s/siirleri/' % (self.url, self.poet),
                             self.parse_all)

    def parse_all(self, response):
        for i in range(1, len(response.css('.pagination li')) - 1):
            yield scrapy.Request('%s/%s/siirleri/ara-/sirala-/sayfa-%s/' %
                                 (self.url, self.poet, i), self.parse_list)

    def parse_list(self, response):
        for i in response.css('.poemListBox a::attr(href)').extract():
            yield scrapy.Request(self.url + i, self.parse)

    def parse(self, response):
        with open('poets/%s.txt' % self.poet, 'a') as f:
            text = ''.join(response.css('.pd-text p::text').extract())
            f.write(text)
            yield {'poem': text}


assignment-3/lstm_text_generation.py
from keras.models import Sequential
from keras.layers import Dense, Activation, LSTM, Dropout, SimpleRNN, GRU
from keras.callbacks import ModelCheckpoint

import numpy as np
import random
import pickle
import os


class TextGenerator:
    def __init__(self, seq_len=50, temperature=0.8, epoch=50, rnn_layer=LSTM):
        self.seq_len = seq_len
        self.temperature = temperature
        self.epochs = epoch
        self.rnn_layer = rnn_layer

    def read_data(self, files_paths):
        self.text = '\n'.join(open(path).read()
                              for path in files_paths).lower()
        self.set_vocab_mapping(self.text)
        return self.text_vec(self.text)

    def text_vec(self, text):
        sentences = []
        next_chars = []
        for i in range(0, len(text) - self.seq_len):
            sentences.append(text[i:i + self.seq_len])
            next_chars.append(text[i + self.seq_len])

        X = np.zeros(
            (len(sentences), self.seq_len, self.vocab_len), dtype=np.bool)
        y = np.zeros((len(sentences), self.vocab_len), dtype=np.bool)
        for i, sentence in enumerate(sentences):
            for t, char in enumerate(sentence):
                X[i, t, self.char_indices[char]] = 1
            y[i, self.char_indices[next_chars[i]]] = 1
        return X, y

    def set_vocab_mapping(self, text):
        chars = sorted(list(set(text)))
        self.vocab_len = len(chars)
        self.char_indices = {c: i for i, c in enumerate(chars)}
        self.indices_char = {i: c for i, c in enumerate(chars)}

    def build_nn(self):
        model = Sequential()
        model.add(
            self.rnn_layer(128, input_shape=(self.seq_len, self.vocab_len)))
        model.add(Dropout(0.2))
        model.add(Dense(self.vocab_len))
        model.add(Activation('softmax'))
        model.compile(loss='categorical_crossentropy', optimizer='adam')
        self.model = model

    def fit(self, X, y):
        self.build_nn()
        filepath = "models/weights-improvement-{epoch:02d}-{loss:.4f}.hdf5"
        checkpoint = ModelCheckpoint(
            filepath,
            monitor='loss',
            verbose=1,
            save_best_only=True,
            mode='min')
        self.model.fit(
            X, y, epochs=self.epochs, batch_size=128, callbacks=[checkpoint])

    def diversity(self, preds):
        preds = np.asarray(preds).astype('float64')
        preds = np.log(preds) / self.temperature
        exp_preds = np.exp(preds)
        preds = exp_preds / np.sum(exp_preds)
        probas = np.random.multinomial(1, preds, 1)
        return np.argmax(probas)

    def predict(self, X):
        return self.model.predict(X, verbose=0)[0]

    def generate(self):
        sentence = self.get_random_sentence()
        generated = list()
        for i in range(400):
            X, _ = self.text_vec(sentence)
            next_index = self.diversity(self.predict(X))
            next_char = self.indices_char[next_index]
            sentence = sentence[1:] + next_char
            generated.append(next_char)
        return ''.join(generated)

    def get_random_sentence(self):
        start_index = random.randint(0, len(self.text) - self.seq_len - 1)
        return self.text[start_index:start_index + self.seq_len + 1]

    def load(self, path):
        self.model.load_weights(path)


# text_generator = TextGenerator(epoch=1)

# X, y = text_generator.read_data('poets/%s' % f for f in os.listdir('poets'))
# text_generator.build_nn()
# text_generator.fit(X, y)

# print(text_generator.generate())

text_generator = TextGenerator(rnn_layer=LSTM, epoch=20)
X, y = text_generator.read_data(['poets/sezai-karakoc.txt'])
text_generator.fit(X, y)

print(text_generator.generate())
